File: convert_missing_webp.js
const sharp = require('sharp');
const fs = require('fs');
const path = require('path');

const missing = [
  'Img/blog/blog-come-scegliere-web-agency',
  'Img/seo-blog-aziendale',
  'Img/aggiornamenti-algoritmo-google-2026',
  'Img/blog/blog-lead-magnet-guida',
  'Img/blog/blog-heatmap-analisi-comportamento',
  'Img/blog/blog-exit-intent-popup',
  'Img/blog/blog-trust-signals-ecommerce',
  'Img/blog/blog-thank-you-page-ottimizzazione',
  'Img/blog/blog-personal-brand-consulente',
  'Img/blog/blog-social-commerce-guida',
  'Img/blog/blog-influencer-marketing-pmi',
  'Img/blog/blog-social-media-automation-tool',
  'Img/blog/blog-tendenze-social-media-2026',
  'Img/blog/blog-bounce-rate-ridurre',
  'Img/blog/blog-data-driven-marketing',
  'Img/blog/blog-marketing-plan-modello',
  'Img/blog/blog-customer-journey-mapping',
  'Img/blog/blog-unique-selling-proposition',
  'Img/blog/blog-omnichannel-strategia',
  'Img/blog/blog-growth-hacking-pmi',
  'Img/blog/blog-competitor-analysis-metodo',
  'Img/blog/blog-digital-transformation-pmi',
  'Img/blog/blog-outsourcing-vs-inhouse-marketing',
  'Img/blog/blog-retention-marketing-strategie',
  'Img/blog/blog-marketing-digitale-attivita-locali',
  'Img/blog/blog-email-vs-sms-marketing',
  'Img/blog/blog-sito-intelligenza-artificiale-limiti',
  'Img/blog/blog-tradurre-sito-web-guida',
  'Img/blog/blog-url-strane-search-console',
  'Img/blog/blog-engagement-sito-web-strategie',
  'Img/blog/blog-dati-obbligatori-sito-web',
  'Img/blog/blog-alternativa-linktree-link-in-bio',
  'Img/blog/blog-analisi-competitiva-online',
];

async function main() {
  let converted = 0, failed = 0, skipped = 0;

  for (const base of missing) {
    const webpPath = base + '.webp';
    
    if (fs.existsSync(webpPath)) {
      skipped++;
      continue;
    }

    // Find source PNG or JPG
    let srcPath = null;
    if (fs.existsSync(base + '.png')) srcPath = base + '.png';
    else if (fs.existsSync(base + '.jpg')) srcPath = base + '.jpg';
    else if (fs.existsSync(base + '.jpeg')) srcPath = base + '.jpeg';

    if (!srcPath) {
      console.log(`❌ NO SOURCE: ${base} (no png/jpg found)`);
      failed++;
      continue;
    }

    try {
      const srcStats = fs.statSync(srcPath);
      await sharp(srcPath)
        .webp({ quality: 85, effort: 6, smartSubsample: true })
        .toFile(webpPath);

      const webpStats = fs.statSync(webpPath);
      const savings = ((1 - webpStats.size / srcStats.size) * 100).toFixed(1);
      console.log(`✅ ${path.basename(srcPath)} → ${path.basename(webpPath)} (${(srcStats.size/1024).toFixed(0)}KB → ${(webpStats.size/1024).toFixed(0)}KB, -${savings}%)`);
      converted++;
    } catch (err) {
      console.log(`❌ ERROR: ${base}: ${err.message}`);
      failed++;
    }
  }

  console.log(`\nDone: ${converted} converted, ${skipped} already existed, ${failed} failed`);
}

main();

File: convert_missing_webp.py
import subprocess, os, re

with open('blog/index.html', 'r', encoding='utf-8') as f:
    content = f.read()

# Find all <source srcset="...webp"> references
sources = re.findall(r'<source srcset="([^"]+\.webp)"', content)

converted = 0
skipped = 0
failed = 0

for src in sources:
    webp_path = src.replace('../', '')
    if os.path.exists(webp_path):
        skipped += 1
        continue
    
    # Find corresponding PNG
    png_path = webp_path.replace('.webp', '.png')
    if not os.path.exists(png_path):
        # Try jpg
        jpg_path = webp_path.replace('.webp', '.jpg')
        jpeg_path = webp_path.replace('.webp', '.jpeg')
        if os.path.exists(jpg_path):
            png_path = jpg_path
        elif os.path.exists(jpeg_path):
            png_path = jpeg_path
        else:
            print(f"❌ NO SOURCE: {webp_path} (no png/jpg found)")
            failed += 1
            continue
    
    # Convert using sips
    try:
        result = subprocess.run(
            ['sips', '-s', 'format', 'webp', '-s', 'formatOptions', '85', png_path, '--out', webp_path],
            capture_output=True, text=True, timeout=30
        )
        if result.returncode == 0 and os.path.exists(webp_path):
            png_size = os.path.getsize(png_path) / 1024
            webp_size = os.path.getsize(webp_path) / 1024
            savings = (1 - webp_size/png_size) * 100 if png_size > 0 else 0
            print(f"✅ {webp_path} ({png_size:.0f}KB → {webp_size:.0f}KB, -{savings:.0f}%)")
            converted += 1
        else:
            print(f"❌ SIPS ERROR: {webp_path}: {result.stderr.strip()}")
            failed += 1
    except Exception as e:
        print(f"❌ EXCEPTION: {webp_path}: {e}")
        failed += 1

print(f"\nDone: {converted} converted, {skipped} already existed, {failed} failed")
